[FEATURE] Search dynamic mode config

rrahn opened the pull request:
Resolves #1436

joergi-w commented on test/unit/search/search_test.cpp:
```diff
@@ -300,6 +300,13 @@ TYPED_TEST(search_test, search_strategy_all)
     }
+
+    { // Test with dynamic mode.
+        mode dynamic_mode{};
```
If I remember correctly, in tests we now have to specify `seqan3::` namespace before all types and functions of seqan3?

rrahn replied:
yes, but I won't have this refactoring done for this PR, since it is about a feature. The other thing must be done separately.

joergi-w commented on include/seqan3/search/configuration/mode.hpp:
```diff
@@ -75,28 +78,166 @@ struct strata : detail::strong_type<uint8_t, strata, detail::strong_type_skill::
+    requires std::same_as<search_mode_t, search_mode_all> ||
+             std::same_as<search_mode_t, search_mode_all_best> ||
+             std::same_as<search_mode_t, search_mode_best> ||
+             std::same_as<search_mode_t, search_cfg::strata>
```
This is already the third occurrence of a check for any valid search mode in this file. Maybe it already makes sense to have a local shortcut like
```cpp
template <typename search_mode_t> 
bool is_search_mode = ...
```

rrahn commented:
@joergi-w I had to make a force push after I rebased onto master. So I also incorporated the changes already into the respective commits. I am sorry if this causes some inconvenience.

joergi-w commented on test/snippet/search/configuration_modes_dynamic.cpp:
```diff
@@ -0,0 +1,19 @@
+    if (true)
+        my_mode = seqan3::search_cfg::best; // The best mode is selected.
+    else
+        my_mode = seqan3::search_cfg::strata{2}; // The strata mode is selected.
```
The `else` branch is never executed...?

rrahn replied:
yes, that is right. but that does not matter. It is a demo to show how to use it. And usually it would be by some condition you would do the one or the other thing. that is also more consistent with the text that describes this snippet.

joergi-w replied:
Okay, fine then 👍

rrahn commented:
@smehringer polite ping

smehringer commented:
I did not manage to review this last week and since this is not an urgent issue and we need to redesign the config, doesn't it make more since to refactor first and then apply this change? Otherwise we need to refactor this code again?

rrahn commented:
> I did not manage to review this last week and since this is not an urgent issue and we need to redesign the config, doesn't it make more since to refactor first and then apply this change? Otherwise we need to refactor this code again?

Since it is not clear when we are going to refactor it, this should be already added. The other point is that the mechanics will be the same, just different class names are used. So I don't see that this can't be incrementally adapted later?

smehringer commented:
I thought we decided to postpone this altogether?

rrahn commented:
> I thought we decided to postpone this altogether?

I think this is a very independent PR. The search result range should wait until after the release because it involves more stuff. But anyway, we could also put that in after the release. I don't care. In general, I was trying to make a point why we should not postpone it until we start refactoring the things. I would like to do it first, when we arrive at an actual use case. But that is independent of the mechanics of this PR.

rrahn commented:
Will be tackled as part of the next iteration with the refined configuration naming.
